### ISSUE TITLE
New-PSSession hangs

### ISSUE THREAD

**Issue by blueleader07** [daverickdunn/full-powershell#5]
First time I'm trying this library (thanks in advance!).  

The first few commands work.  But then it hangs when I try to open a session.  No response ever comes back in the subscribe handler.

I'm wrapping the call method in a promise so that I can use await and run synchronously.  Perhaps that's a bad idea...?
```
export const call = async (shell: any, command: string, format?: Format) : Promise<any> => {
    return new Promise((resolve, reject) => {
        shell.call(command, format).subscribe(
            (response: any) => {
                resolve(response.success)
            },
            (error: any) => {
                reject(error)
            }
        )
    })
}
```

These commands all work ... setting up the credentials ... 

```
export const getCredential = async (ps: any) => {
    await call(ps, `$User = "${environmentUtils.getVariable('POWERSHELL_USER')}"`, 'string')
    await call(ps, `$PWord = ConvertTo-SecureString -String "${environmentUtils.getVariable('POWERSHELL_PASSWORD')}" -AsPlainText -Force`, 'string')
    await call(ps, '$UserCredential = New-Object -TypeName System.Management.Automation.PSCredential -ArgumentList $User, $PWord', 'string')
}
```

Then it hangs when I do this ... 
```
export const getCredential = async (ps: any) => {
    await call(ps, '$SccSession = New-PSSession -ConfigurationName Microsoft.Exchange -ConnectionUri https://ps.compliance.protection.outlook.com/powershell-liveid/ -Credential $UserCredential -Authentication "Basic" -AllowRedirection', null)
    await call(ps, 'Import-PSSession $SccSession -Prefix cc', null)
    await call(ps, '$Session = New-PSSession -ConfigurationName Microsoft.Exchange -ConnectionUri https://ps.compliance.protection.outlook.com/powershell-liveid/ -Credential $UserCredential -Authentication Basic –AllowRedirection', null)
    await call(ps, 'Import-PSSession $Session', null)
}
```

This all works with the node-powershell library.  Any ideas why it doesn't response with full-powershell?  Perhaps I am doing something wrong.

**Comment by daverickdunn** [comment]
The promise is fine. The variables appear to be lost across calls. I've been generally passing it scripts with one output, or calling simple commands. I've not really tried building up state between calls. It should of course be able to do that, it's just not the way I've written my code. I'll take a look over the weekend, I suspect it won't be too hard to fix.

**Comment by daverickdunn** [comment]
@blueleader07 So the problem is that I wrap whatever commands you pass in another PowerShell function that allows me to serialise the output streams and catch errors, etc. This scopes your variables to that function, and thus they're not in the same scope between calls.

For now, you can prefix your variable with `$global:` as a workaround. I'll try figure out a permanent solution later.

So, in your case, it'd be:

```typescript
export const getCredential = async (ps: any) => {
    await call(ps, `$global:User = "${environmentUtils.getVariable('POWERSHELL_USER')}"`, 'string')
    await call(ps, `$global:PWord = ConvertTo-SecureString -String "${environmentUtils.getVariable('POWERSHELL_PASSWORD')}" -AsPlainText -Force`, 'string')
    await call(ps, '$UserCredential = New-Object -TypeName System.Management.Automation.PSCredential -ArgumentList $User, $PWord', 'string')
}
```

```typescript
export const getCredential = async (ps: any) => {
    await call(ps, '$global:SccSession = New-PSSession -ConfigurationName Microsoft.Exchange -ConnectionUri https://ps.compliance.protection.outlook.com/powershell-liveid/ -Credential $UserCredential -Authentication "Basic" -AllowRedirection', null)
    await call(ps, 'Import-PSSession $SccSession -Prefix cc', null)
    await call(ps, '$global:Session = New-PSSession -ConfigurationName Microsoft.Exchange -ConnectionUri https://ps.compliance.protection.outlook.com/powershell-liveid/ -Credential $UserCredential -Authentication Basic –AllowRedirection', null)
    await call(ps, 'Import-PSSession $Session', null)
}
```

**Comment by daverickdunn** [comment]
Fixed with: 197ea63
Released as: [v1.2.7](https://www.npmjs.com/package/full-powershell/v/1.2.7)

**Comment by daverickdunn** [comment]
And just to add, [1.3.0](https://www.npmjs.com/package/full-powershell/v/1.3.0) now exposes [promises](https://github.com/daverickdunn/full-powershell/blob/master/README.md#promises).

**Comment by blueleader07** [comment]
@daverickdunn first, thank you so much for the quick fix.  I confirmed that it works as expected locally, which is exciting!  Now I'm trying to deploy it (as a lambda) and running into to odd behavior.  I've been working on node 12.18.3 (probably time to update).  Just curious ... what node version are you developing on?

**Comment by daverickdunn** [comment]
No problem. It was a glaring issue that I managed not to come across because I tend to read and execute entire `.ps1` files, rather than break down my statements into individual calls.

It's definitely been run on 12.x but I'm pretty sure it's also had plenty of execution time on 13.x and 14.x without any problems.

The flavour of PowerShell itself is much more of a pain-point in my experience. What Lambda runtime are you using? Last I checked there were a lot of features not present in Core that I required.

**Comment by daverickdunn** [comment]
I should probably add that part of the motivation for writing this was package, besides the lockup issues in node-powershell, was to address the random output from remote powershell calls. They are often, irregular, and they change, so it was hard to parse that noise all from one stream. I've had a few debates on the PowerShell GitHub issues page about additional messages that just get written to stdout by the underlying .net libs, but they don't seem to care. I think what I've done here is about as clean as you can make it without using C#, etc.

**Comment by blueleader07** [comment]
I'm running Node 12.x on Lambda.  I have a powershell "layer" that I created to add the additional dependencies.  This works with "node-powershell" ... but it hangs on invoke() sometimes.  Now I'm attempted to run full-powershell in the same environment ... everything it needs from powershell should be there.

Here's how my full-powershell output looks when I run it locally (opening the connection to Microsoft):

```
connecting powershell ... 
output {
  success: [],
  error: [],
  warning: [],
  verbose: [],
  debug: [],
  info: []
}
output {
  success: [],
  error: [],
  warning: [],
  verbose: [],
  debug: [],
  info: []
}
output {
  success: [],
  error: [],
  warning: [],
  verbose: [],
  debug: [],
  info: []
}
output {
  success: [],
  error: [],
  warning: [
    'Your connection has been redirected to the following URI: "https://nam12b.ps.compliance.protection.outlook.com/powershell-liveid?PSVersion=7.1.3\u0000"\n'
  ],
  verbose: [],
  debug: [],
  info: []
}
output {
  success: [
    '\n' +
      'ModuleType Version    PreRelease Name                                ExportedCom\n' +
      '                                                                     mands\n' +
      '---------- -------    ---------- ----                                -----------\n' +
      'Script     1.0                   tmp_zcl0bmqb.u0x                    {Add-ccCom…\n' +
      '\n'
  ],
  error: [],
  warning: [],
  verbose: [],
  debug: [],
  info: []
}
output {
  success: [],
  error: [],
  warning: [
    'Your connection has been redirected to the following URI: "https://nam12b.ps.compliance.protection.outlook.com/powershell-liveid?PSVersion=7.1.3\u0000"\n'
  ],
  verbose: [],
  debug: [],
  info: []
}
output {
  success: [
    '\n' +
      'ModuleType Version    PreRelease Name                                ExportedCom\n' +
      '                                                                     mands\n' +
      '---------- -------    ---------- ----                                -----------\n' +
      'Script     1.0                   tmp_fzdiwiwd.bl3                    {Add-Compl…\n' +
      '\n'
  ],
  error: [],
  warning: [],
  verbose: [],
  debug: [],
  info: []
}
connected in 00:00:48

```

As you can see it took 48 seconds to open the powershell session (this is expected and the same behavior as node-powershell).  

Here's how it looks when run from lambda:

```
2021-07-10T12:35:11.432Z	6cff3302-56e9-49b5-ad8b-801d9822b432	INFO	connecting powershell ... 
2021-07-10T12:35:14.129Z	6cff3302-56e9-49b5-ad8b-801d9822b432	INFO	output {
  success: [ '' ],
  error: [ '' ],
  warning: [ '' ],
  verbose: [],
  debug: [],
  info: [ '' ]
}
2021-07-10T12:35:14.233Z	6cff3302-56e9-49b5-ad8b-801d9822b432	INFO	output {
  success: [ '' ],
  error: [ '' ],
  warning: [ '' ],
  verbose: [],
  debug: [],
  info: [ '' ]
}	
2021-07-10T12:35:14.382Z	6cff3302-56e9-49b5-ad8b-801d9822b432	INFO	output {
  success: [ '' ],
  error: [ '' ],
  warning: [ '' ],
  verbose: [],
  debug: [],
  info: [ '' ]
}
2021-07-10T12:35:14.746Z	6cff3302-56e9-49b5-ad8b-801d9822b432	INFO	output {
  success: [ '' ],
  error: [ '' ],
  warning: [ '' ],
  verbose: [],
  debug: [],
  info: [ '' ]
}
2021-07-10T12:35:14.882Z	6cff3302-56e9-49b5-ad8b-801d9822b432	INFO	output {
  success: [ '' ],
  error: [ '' ],
  warning: [ '' ],
  verbose: [],
  debug: [],
  info: [ '' ]
}
2021-07-10T12:35:15.032Z	6cff3302-56e9-49b5-ad8b-801d9822b432	INFO	output {
  success: [ '' ],
  error: [ '' ],
  warning: [ '' ],
  verbose: [],
  debug: [],
  info: [ '' ]
}	
2021-07-10T12:35:15.164Z	6cff3302-56e9-49b5-ad8b-801d9822b432	INFO	output {
  success: [ '' ],
  error: [ '' ],
  warning: [ '' ],
  verbose: [],
  debug: [],
  info: [ '' ]
}
2021-07-10T12:35:15.166Z	6cff3302-56e9-49b5-ad8b-801d9822b432	INFO	connected in 00:00:03
```

Oddly, output seems to return almost immediately and with only empty strings.  Obviously it's not "really" connecting in only 3 seconds.  

So, I suspect it has something to do with RxJS and lambda.  I tried with your call.promise() and with my own call() wrapper (from my earlier messages) and got the same behavior.  I'm using global$: for my variables and that works fine locally (not sure if the behavior is different in lambda). 

```
export const getCredential = async (ps: any) => {
    await runCommand(ps, `$global:User = "${environmentUtils.getVariable('POWERSHELL_USER')}"`)
    await runCommand(ps, `$global:PWord = ConvertTo-SecureString -String "${environmentUtils.getVariable('POWERSHELL_PASSWORD')}" -AsPlainText -Force`)
    await runCommand(ps, '$global:UserCredential = New-Object -TypeName System.Management.Automation.PSCredential -ArgumentList $global:User, $global:PWord')
}
``` 

```
    async connect () {
        console.log('connecting powershell ... ')
        const ps = new PowerShell()
        await getCredential(ps)
        await runCommand(ps, '$global:SccSession = New-PSSession -ConfigurationName Microsoft.Exchange -ConnectionUri https://ps.compliance.protection.outlook.com/powershell-liveid/ -Credential $global:UserCredential -Authentication "Basic" -AllowRedirection', false)
        await runCommand(ps, 'Import-PSSession $global:SccSession -Prefix cc', false)
        await runCommand(ps, '$global:Session = New-PSSession -ConfigurationName Microsoft.Exchange -ConnectionUri https://ps.compliance.protection.outlook.com/powershell-liveid/ -Credential $global:UserCredential -Authentication Basic –AllowRedirection', false)
        await runCommand(ps, 'Import-PSSession $global:Session', false)
        return ps
    },
```

So that's where I'm at.  So close!  I'll keep plugging away and post my results here incase anyone else runs into the same thing.  It's probably my tsconfig build or something and nothing to do with full-powershell.

**Comment by blueleader07** [comment]
For the record, here's my .tsconfig
```
{
  "compilerOptions": {
    "allowJs": true,
    "experimentalDecorators": true,
    "strictPropertyInitialization":false,
    "target": "ES2019",
    "module": "commonjs",
    "strict": true,
    "esModuleInterop": true,
    "skipLibCheck": true,
    "forceConsistentCasingInFileNames": true,
    "outDir": "build",
    "rootDir": "./src",
    "lib": [
      "es2019",
      "dom"
    ]
  },
  "include": [
    "src/**/*.ts"
  ]
}

```

**Comment by daverickdunn** [comment]
I see, that's very interesting re: the PowerShell layer. I would love to know more about how you created this as our biggest infra cost at the moment is running a Windows Server ec2 cluster.

What version of Pwsh is it?

Straight away I can see a potential issue using promises. The way this library works is that only the _first_ emission is returned when using the Promise returned from `.call()`, or any promise for that matter. So if the warning you get from Exchange is emitted before the session has been created, your next call is going to happen too early. Also worth noting that this being an async problem it may be intermittent, so could explain different results at different times.

Verbose and Debug were the first potential sources of problems I thought of when you mentioned Lambda, as they require writing to disk. I believe Lambda does have a `/tmp` folder available, but I'm not sure what happens when the current directory doesn't have write access. I'll have to test this as it could very well be causing problems.

It's strange that the second run returns empty strings for most of the streams but the first doesn't. Both emitted 7 times, so that maybe indicates that they both ran the commands. I'll have a think about why that might be. It doesn't happen for Verbose and Debug.

I strongly doubt there's an issue with RxJS. It's a very popular and well defined lib; Angular is built around it. That said, I had a think about how to pipe the RxJS streams to better control the async flow, and while it's completely fine for someone who's used to RxJS, it is a bit verbose and requires a bit of experience/learning. I have an idea on how to simplify this, gonna play around with it a little and see how it goes, will probably have an update later.

Which version of TypeScript are you using?

**Comment by daverickdunn** [comment]
Had a closer look this morning and I'm pretty sure the way I capture the Verbose/Debug streams is indeed the issue. Classic error handling no-no. I vaguely remember thinking I had better write this part better, always the way...

```powershell
    $vv = "fps_verbose.tmp"
    $dv = "fps_debug.tmp"
    try {
        Invoke-Command -NoNewScope -ScriptBlock {
            ${command}
        } 1>$null 2>$null 3>$null 4>$vv 5>$dv 6>$null -ov ov -ev ev -wv wv -iv iv
    } 
    catch { }
    finally {
        $verbose = Get-Content $vv
        $debug = Get-Content $dv
        # return all streams ...
    }
```

PowerShell redirects are not all alike, and the Verbose and Debug streams cannot be redirected to variables like the other four. But I figured out that you can redirect them to files and then read the file back afterwards. That all works fine when the file can be created, but I never tested what happens when it can't, as is the case with Lambda. The part I was meant to fix was the catch block, because as you can see, it'll silently suppresses script errors arising from the `Invoke-Command` failing. It's the `Invoke-Command` that tries to create the files.

I need to figure out how to test this failure and then provide an option to either change the path to the `.tmp` files, or an option to disable those streams entirely.

**Comment by daverickdunn** [comment]
Please see the fix discussed in issue #8
